### ISSUE TITLE
[Feature Request] Formally validate public API modification between releases

### ISSUE THREAD

**Issue by mjameswh** [temporalio/sdk-typescript#938]
### Is your feature request related to a problem? Please describe.

At the moment, it is very difficult to identify changes to exported members signatures from our SDK packages. This leads to two negative consequences:

- Risk of unexpected breaking changes in patch or minor releases;
- Risk of exposing members that were never meant for public usage (thus increasing even more the risk of further breaking changes).

### Describe the solution you'd like

Use a tool run in CICD that:
1. Identify the exact set of publicly exported APIs
2. Reports changes in these exported APIs relative to the appropriate baseline

### Additional context

Microsoft's [API Extractor](https://api-extractor.com/) can probably be used for this.

**Comment by bergundy** [comment]
Good idea, I'd just note that we're reserving major versions for history incompatible changes, our minor releases may contain API breaking changes.
